### PR TITLE
Use ASM to invoke event handlers

### PR DESCRIPTION
**Pull request by mikroskeem**

### CODE REVIEW

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
++        {
++            MethodVisitor mv = cw.visitMethod(Opcodes.ACC_PUBLIC, "execute", eeDesc, null, null);
++            mv.visitVarInsn(Opcodes.ALOAD, 1);
++            mv.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(listenerClass));
++            mv.visitVarInsn(Opcodes.ALOAD, 2);
++            mv.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(eventClass));
++            int inv = listenerClass.isInterface() ? Opcodes.INVOKEINTERFACE : Opcodes.INVOKEVIRTUAL;
++            mv.visitMethodInsn(inv, Type.getInternalName(listenerClass), handlerMethod.getName(), Type.getMethodDescriptor(handlerMethod), listenerClass.isInterface());
++            if(handlerMethod.getReturnType() != void.class) {
++                mv.visitInsn(Opcodes.POP);
```
Comment (Janmm14) [inline BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: needs pop2 for long/double

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
++import java.lang.reflect.Modifier;
++
++/**
++ * @author Mark Vainomaa
++ */
++public interface EventExecutor {
++    void execute(Object listener, Object event);
++
++    static EventExecutor create(Method method) {
++        if (Modifier.isPublic(method.getModifiers())) {
```
Comment (Janmm14) [inline BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: does magicaccessor class still exist? if yes, we could do nonpublic calls also with asm

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
++
++import java.lang.invoke.MethodHandle;
++import java.lang.invoke.MethodHandles;
++import java.lang.reflect.Method;
++import java.lang.reflect.Modifier;
++
++/**
++ * @author Mark Vainomaa
++ */
++class MethodHandleEventExecutor implements EventExecutor {
```
Comment (Janmm14) [inline BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: i think there were huge discussions about method handle vs normal reflection and i heard that methodhandles are in the end slower than normal reflection, as the jvm optimizes normal reflection also to asm at some point

**Comment by Janmm14** [comment]
This was part of waterfall in the past, someone should dig up removal reasons.

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch (hunk 2 above)
Comment (mikroskeem) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: Can't access magic stuff past Java 9 iirc, wouldn't risk with that.

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch (hunk 3 above)
Comment (mikroskeem) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: Yeeah maybe indeed.

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch (hunk 2 above)
Comment (Janmm14) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: Actually I don't think the Modifier will ever be nonpublic, as on register it only registers public methods or sth like that

Comment (mikroskeem) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: Will check it later and make changes.

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch (hunk 1 above)
Comment (mikroskeem) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: Does EventBus even consider methods with non-void returns as event handlers? Then could simplify that logic and remove POP entirely.

Comment (Janmm14) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: EventBus does not check return value of `@EventHandler` methods: https://github.com/SpigotMC/BungeeCord/blob/b85df4f2a19222bf96b9143de38438a95911eec4/event/src/main/java/net/md_5/bungee/event/EventBus.java#L61

**Comment by mikroskeem** [comment]
Archaeological excavations resulted with this: <https://github.com/PaperMC/Waterfall/commit/80bed7979d707563283635998cc326f6aa0ad5b0#commitcomment-18806881>

tl; dr; issue with Techcable's Event4J library, which is not an issue here anymore as this is more raw solution.

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch (hunk 1 above)
Comment (mikroskeem) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: Where is the facepalm emoji in GitHub when I need it.

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch (hunk 2 above)
Comment (mikroskeem) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: register calls findHandlers, which in turn calls `listener.getClass().getDeclaredMethods()`, so we still end up with non-public methods as I recalled correctly.

Comment (Janmm14) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: bungee does call getDeclaredMethods but does never call setAccessible(true), so it would fail calling a nonpublic eventhandler

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch (hunk 3 above)
Comment (mikroskeem) [reply BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch]: In case of Bungee not executing non-public methods (either intentionally or not intentionally), then this event executor implementation becomes irrelevant.

**Comment by mikroskeem** [comment]
Made requested changes.

**Comment by Leymooo** [comment]
> This was part of waterfall in the past, someone should dig up removal reasons.

https://github.com/PaperMC/Waterfall/issues/70
https://github.com/PaperMC/Waterfall/issues/84

File: BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+diff --git a/event/src/main/java/net/md_5/bungee/event/EventBus.java b/event/src/main/java/net/md_5/bungee/event/EventBus.java
+index d7c3b90d..2aa4941d 100644
+--- a/event/src/main/java/net/md_5/bungee/event/EventBus.java
++++ b/event/src/main/java/net/md_5/bungee/event/EventBus.java
+@@ -51,10 +51,10 @@ public class EventBus
+                 } catch ( IllegalArgumentException ex )
+                 {
+                     throw new Error( "Method rejected target/argument: " + event, ex );
+-                } catch ( InvocationTargetException ex )
++                } catch ( Exception ex ) // Waterfall - Use ASM to invoke event handlers
```
Comment (mikroskeem) [inline BungeeCord-Patches/0056-Use-ASM-to-invoke-event-handlers.patch:177]: Though this makes me think. What happens when `Method#invoke(...)` throws `Throwable` instead of an `Exception`? Javadoc tells me only about [exception](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Method.html#invoke-java.lang.Object-java.lang.Object...-).

This needs to be verified, if Throwable was passed on wrapped in InvocationTargetException, then I need to change this catch to handle Throwable instead.

**Comment by xxDark** [comment]
There are still some things that can be improved by dropping `CHECKCAST` instructions when possible (see `jdk.internal.reflect.MagicAccessorImpl/sun.reflect.MagicAccessorImpl`)

**Comment by Janmm14** [comment]
why got this closed?
@xxDark I don't think that normal java classes can leave out CHECKCAST, it'd get a verify error likely on class load

**Comment by Hyronymos-Old** [comment]
I guess he closed all his PRs because they didnt got merged after a month and because he is active contributing to velocity now

**Comment by mikroskeem** [comment]
>why got this closed?

Low interest in implementing this. If someone else wants to do this, feel free to pick up the leftovers.

**Comment by xxDark** [comment]
> 
> 
> why got this closed?
> @xxDark I don't think that normal java classes can leave out CHECKCAST, it'd get a verify error likely on class load

Doing that without any problems. Plus it allows me to invoke private listener's methods
